app/functoria_app.ml

```diff
@@ -417,81 +435,113 @@ module Make (P: S) = struct
     ()
 
   let clean_main i jobs =
-    Engine.clean i jobs >>| fun () ->
-    Cmd.remove (Info.root i / "main.ml")
+    Engine.clean i jobs >>= fun () ->
+    Bos.OS.File.delete Fpath.(v "main.ml")
 
   let configure i jobs =
-    Log.info "%a %s" Log.blue "Using configuration:"  (get_config_file ());
-    Log.info "opam: %a" (Info.opam ?name:None) i ;
-    Cmd.in_dir (Info.root i) (fun () -> configure_main i jobs)
+    Log.info (fun m -> m "Using configuration: %s" (get_config_file ()));
+    Log.info (fun m -> m "opam: %a" (Info.opam ?name:None) i);
+    Log.info (fun m -> m "within: %s" (Info.root i));
+    match
+      Bos.OS.Dir.with_current (Fpath.v (Info.root i))
+        (fun () -> configure_main i jobs) ()
+    with
+    | Ok a -> a
+    | Error _ -> R.error_msg "failed to change directory for configure"
+
+  let build i jobs =
+    Log.info (fun m -> m "Building: %s" (get_config_file ()));
+    match
+      Bos.OS.Dir.with_current (Fpath.v (Info.root i))
+        (fun () -> Engine.build i jobs) ()
+    with
+    | Ok a -> a
+    | Error _ -> R.error_msg "failed to change directory for build"
 
   let clean i (_init, job) =
-    Log.info "%a %s" Log.blue "Clean:"  (get_config_file ());
+    Log.info (fun m -> m "Cleaning: %s" (get_config_file ()));
     let root = Info.root i in
-    Cmd.in_dir root (fun () ->
-        clean_main i job >>= fun () ->
-        Cmd.run "rm -rf %s/_build" root >>= fun () ->
-        Cmd.run "rm -rf log %s/main.native.o %s/main.native %s/*~" root
-          root root
-      )
+    match
+      Bos.OS.Dir.with_current (Fpath.v root)
+        (fun () ->
+           clean_main i job >>= fun () ->
+           Bos.OS.Dir.delete ~recurse:true (Fpath.v "_build") >>= fun () ->
+           Bos.OS.File.delete (Fpath.v "log") >>= fun () ->
+           Bos.OS.File.delete (Fpath.v "main.native.o") >>= fun () ->
+           Bos.OS.File.delete (Fpath.v "main.native")) ()
+    with
+    | Ok a -> a
+    | Error _ -> R.error_msg "failed to change directory for configure"
 
   (* FIXME: describe init *)
   let describe _info ~dotcmd ~dot ~output (_init, job) =
     let f fmt = (if dot then Config.pp_dot else Config.pp) fmt job in
     let with_fmt f = match output with
       | None when dot ->
-        let f oc = Cmd.with_channel oc f in
-        Cmd.with_process_out dotcmd f
-      | None -> f Fmt.stdout
-      | Some s -> Cmd.with_file s f
-    in R.ok @@ with_fmt f
+        f Format.str_formatter ;
+        let data = Format.flush_str_formatter () in
+        Bos.OS.Cmd.run_in (Bos.Cmd.v dotcmd) (Bos.OS.Cmd.in_string data)
```

Comment:
While you're at it (and you seem fluent in `Bos`), could you pipe that to a temp file and call `dotcmd` on the tempfile, instead ? See #81.